### ISSUE TITLE
setup credit and deduction system

### ISSUE THREAD

**Issue by balupton** [bevry-archive/sponsored#11]
as the only reliable data we can ascertain from services/sources is historical data - that relate to cumulative donation amounts - i.e. credit - rather than current month/tier details, instead we need to implement a system that does rewards by deductions from determined credit instead

The question then becomes how should deductions work?

There are cycle timing options:

- daily: start of next day to end of next day, according to some time zone
- monthly: start of next day, to the same date on the following month, according to some time zone
- monthly: start of next month, to the end of next month, according to GMT - this is what patreon does, but means if you donate on 2nd of March, you gotta wait until 1st of April for your sponsorship to go through (perhaps we can have an override button to force this month to be the sponsored month)

There are extra donation considerations:

- if a user has donated extra money than what their current selected tier is, then should that money go to the next month,
- or should that credit be consumed in levelling up their current month level up (perhaps could be a override button)

There are specification options:

- deductions occur by automatically consuming the highest tier that the credit can occur - this could consume credits quicker than expected, as a higher tier could be consumed over one month, when the plan could have been to spread out a lower tier over multiple months
- deductions occur by whichever tier the user has specified

There are notification options:

- when the user runs out of credit for the next cycle, they are notified to adjust their funding service
- if sources (stripe, paypal, coinbase) are configured, they could be billed automatically until they cancel
- if sources are configured, they could get an email to manually top up their credit
- when a users contribution on a funding service has determined to be ended, we send a thank you email to them for the time they did spend, and an email summary of all that they funded, and the exposure they got

Another option for the extra donation, and timing things can be:

- where if someone donates on the 14th of March, but doesn't want to wait until the 1st of April for their sponsorship to show, they could:
  - pro-rata it, so the deduction is only for 50% of the month - this makes historical listings complicated, as they effectively could get higher tiers at super discounted prices inside historical rendering
  - override it, so the deduction occurs like normal at 100% of the tier for the month, and they just sacrifice the passed days
  - or where such situations are not allowed, and they have to wait until the next month
  - the day cycle, and the date to date cycle, avoids this issue, as they just have to wait a day at most for the sponsorship to kick in - but it complicates historical listings as I'm not sure how you would render those things

One potential for rendering month to month historical listings (donated amounts will not be listed, just there for info)

```
# Backers

## 2017

### 03 March

- Bob Bill with url and description (donated $XX amount over 3 months, from 2017-01 to 2017-03)

### 02 February

- Jim Jones with url and description and logo (donated $XXX amount over 2 months, from 2017-01 to 2017-02)
- Bob Bill with url and description (donated $XX amount over 3 months, from 2017-01 to 2017-03)

### 01 January

- Jim Jones with url and description and logo (donated $XXX amount over 2 months, from 2017-01 to 2017-02)
- Bob Bill with url and description (donated $XX amount over 3 months, from 2017-01 to 2017-03)
- Richard Richie with url (donated $X amount over 1 month, on 2017-01)
```

One potential for rendering historical listings for all of the options, including date to date and day to day, would be the exact same method for rendering the current months listing, but where we pull everyone in:

```
# Backers

- Jim Jones with url and description and logo (donated $XXX amount over 2 months, from 2017-01 to 2017-02)
- Bob Bill with url and description and logo (donated $XXX amount over 3 months, from 2017-01 to 2017-03)
- Richard Richie with url (donated $X amount over 1 month, on 2017-01)
```

Notice how bob bill gets promoted, as his three $50/50/$50 listings put in the $150 tier - this is the only way that I think this listing style can properly reward people who have donated over a long period of time.

**Comment by balupton** [comment]
(current sponsors @aglezabad @jlgeering) (historical sponsors @scokem @ahdinosaur) as you fellows will be the benefactors of the above system, which options take your fancy?

**Comment by balupton** [comment]
going to go monthly option, as it can be converted to the daily option at a later time if necessary, but the daily option cannot be converted to the monthly option

**Comment by Aglezabad** [comment]
I agree with the monthly option and I think it will be easier to do if you are going to use GMT instead of specific time zone.

Sorry for not telling anything before. I was busy.

**Comment by jlgeering** [comment]
monthly sounds fine if the reward is recognition (being listed as a backer), i.e. wait for next month to be listed. Not sure about the small amount over time vs 1 big lump question, but probably the second option with a single aggregate list. Also it avoids repetition.

But then the "fidelity" aspect gets lost (how long has X been a supporter) ... deleted 2 bad proposals how to address that => no clue how to "fix" it, not sure if it even matters.

PS: I'm actually surprised (in a good way) that you noticed me supporting you ... first time I get a confirmation that Gratipay is working ;-)

**Comment by balupton** [comment]
cool, thanks for the above, I will get back to this in the next few days!

@jlgeering while I don't have the code up right now, I believe you were pulled in from your open collective donations! as currently gratipay doesn't expose who our sponsors are

:)

**Comment by balupton** [comment]
Status update.

I've been working on the visual web interface the past few days...

![screen shot 2017-04-05 at 3 49 43 pm](https://cloud.githubusercontent.com/assets/61148/24695241/8659ba32-1a17-11e7-9cb0-bfc1dcf4a87f.png)

I've also gone and coded [a Patreon scraper](https://github.com/bevry/sponsored/commit/22d79975755b19a7f82811265bc8f05f1dd66495) to get the complete donation records of all the backers - not just the current months ones.

This is what I've concluded...

Monthly billing, tier selection, a credit and deduction system, and their displays, are all issues unique to the time period model - and they are all very complex.

E.g. What happens when you change the tiers you offer? Well, you would need to save them for the next billing cycle. As well as move everyone to the closest lower tier that is related to the one they were previously on. As well as notify everyone of the new fields they need to fill out.

They also make it very hard to develop this thing, as unless one adds in an override ability, then all your changes are queued until the next billing cycle. Which slows down the development of this.

As such...

I'm going to abolish the monthly tier model, and just consider it a legacy of the Patreon system. Instead, we Sponsored will assign tiers to the cumulative amount the user has donated. Users who have donated in the past ~31 days will be rendered in READMEs and Widgets, with a link to see the historical backers.

Backers will be grouped by their tier, and backers be sorted with highest cumulative amount first.

This will:

- drastically simplify the complexity of the system's implementation as well as its conceptualisation
- be fairer on sponsors (no delays in their rendering), and consequently require less legal/accounting issues if we ever screw up the complex rendering of time-based billing cycles
- be quicker for developers/admins/maintainers to revise and improve billing tiers and their rendering
- correctly reward people who have given the most, even if it was over a long period of time
- still incentivise active sponsors vs inactive sponsors

It hits all the goals really well.

Let me know your thoughts. This is quite the radical departure from the way these things are currently implemented.
